README.md
# Micropython demo clock on ESP32
ESP32 with DHT22, DS18B20, DS3231 sensors and dot LED matrix on MAX7219

## Download Micropython firmware

Go to:

1. "Getting started with MicroPython on the ESP8266"
http://docs.micropython.org/en/latest/esp8266/esp8266/tutorial/intro.html

2. "Firmware for ESP32 boards"
look http://micropython.org/download

## Install `esptool.py` - ESP8266 & ESP32 ROM Bootloader Utility
```
$ sudo apt-get install python-pip
$ sudo pip install esptool
```
OR
```
$ sudo apt-get install python-serial
$ sudo apt-get install python-ecdsa ecdsautils
$ sudo apt-get install python-slowaes
$ git clone https://github.com/espressif/esptool.git
$ cd esptool
$ sudo python setup.py install
```

## Flash firmware on ESP8266 board
The following files are daily firmware for ESP32-based boards.
Program your board using the esptool.py program, and put the firmware
starting at address 0x1000

If you are putting MicroPython on for the first time then you should
first erase the entire flash. 

```
$ esptool.py --chip esp32 --port /dev/ttyUSB0 erase_flash
$ esptool.py --chip esp32 --port /dev/ttyUSB0 write_flash -z 0x1000 firmware.bin

```

## Build `mpy-cross`

```
$ sudo apt-get install build-essential libreadline-dev libffi-dev git
$ git clone --recurse-submodules https://github.com/micropython/micropython.git
$ cd ./micropython/ports/unix
$ make axtls
$ make
$ sudo cp ./micropython /usr/local/bin
```

## Install `ampy` - Utility to interact with a MicroPython board over a serial connection
```
$ sudo pip install adafruit-ampy
```
OR
```
$ sudo apt-get install python-click
$ git clone https://github.com/adafruit/ampy.git
$ cd ampy
$ sudo python setup.py install
```

## Compile python modules
```
$ mpy-cross main.py
$ mpy-cross ds3231.py
$ mpy-cross max7219.py
$ mpy-cross font_4x6.py
```

## Download files to ESP32 FLASH
```
$ ampy --port /dev/ttyUSB0 put main.py
$ ampy --port /dev/ttyUSB0 put ds3231.mpy
$ ampy --port /dev/ttyUSB0 put max7219.mpy
$ ampy --port /dev/ttyUSB0 put font_4x6.mpy
```

## Connect sensors and display to ESP32

Connect DTH22 to GND, 3.3V and GPIO-15

Connect DS18B20 to GND, 3.3V and GPIO-4

Connect DS3231 to GND, 3.3V, D21 (GPIO-21) to SDA and D22 (GPIO-22) to SCL

Connect MAX7219 to SPI:

| My ESP32   | ESP32         | max7219 8x8x4 LED Matrix |
| ---------- | ------------- | ------------------------ |
| 3V         | 3.3V          | VCC                      |
| GND        | GND           | GND                      |
| D14        | GPIO14 = SCK  | CLK                      |
| D12        | GPIO12 = MISO | n/c                      |
| D13        | GPIO13 = MOSI | DIN                      |
| D27        | GPIO27        | CS (software)            |


## Run terminal (minicom, picocom or screen)
```
$ minicom -D /dev/ttyUSB0 -b 115200

$ picocom /dev/ttyUSB0 -b 115200

$ screen /dev/ttyUSB0 115200
```

## Set date/time in terminal
```
>>> rtc.set(datetime=(2017, 12, 31, 0, 23, 59, 00)
```

## Be happy!

[youtube video link](https://www.youtube.com/watch?v=WTeM4eQrp3c)



ChangeLog
2018.02.01
 + first real test and initial commit
 * change 4x6 pixels font to 6x8

2018.02.04
 * fix bug in weekday index



main.py
# Demo clock on ESP32
import network
import socket
from machine import Pin, SPI, Timer
import time
from dht import DHT22
import onewire
import ds18x20
from ds3231 import *
import max7219
#from font_4x6 import font_4x6 as FONT
from font_6x8 import font_6x8 as FONT
import framebuf
import gc
gc.collect()

# setup Wi-Fi network
sta_if = network.WLAN(network.STA_IF)
sta_ap = network.WLAN(network.AP_IF)

"""
if False:
    # create AP
    sta_if.active(False)
    sta_ap.active(True)
    sta_ap.config(essid="esp-net", authmode=network.AUTH_WPA_WPA2_PSK, password="radio86rk")
    # IP address, netmask, gateway, DNS
    sta_ap.ifconfig(['192.168.1.1', '255.255.255.0', '192.168.1.1', '8.8.8.8'])

else:
    # connnect to AP
    sta_ap.active(False)
    sta_if.active(True)
    sta_if.connect('veonet', 'radio86rk')
    sta_if.ifconfig(['192.168.0.12', '255.255.255.0', '192.168.0.254', '192.168.0.254'])
    #sta_if.ifconfig()
"""

# DHT22 wrapper
class _DHT22_:
    def __init__(self, gpio_n):
        self.sensor = DHT22(Pin(gpio_n, Pin.IN, Pin.PULL_UP))
        self.t = 0
        self.h = 0
        self.m = True

    def run(self):
        if self.m:
            self.m = False
            self.sensor.measure() # poll DHT22 sensor
        else:
            self.m = True
            self.t = self.sensor.temperature()
            self.h = self.sensor.humidity()
            #print("DHT22: t=%f, h=%f" % (self.t, self.h))

    def get(self):
        return self.t, self.h

# DS18B22 wrapper
class _DS18B20_:
    def __init__(self, gpio_n):
        self.ow = onewire.OneWire(Pin(gpio_n)) # create a OneWire bus on pin
        self.ow.scan()  # return a list of devices on the bus
        self.ow.reset() # reset the bus
        self.ds = ds18x20.DS18X20(self.ow)
        self.t = 0
        self.m = True

    def run(self):
        if self.m:
            self.m = False
            self.roms = self.ds.scan()
            self.ds.convert_temp()
        else:
            self.m = True
            for rom in self.roms:
                self.t = self.ds.read_temp(rom)
                #print("DS18B20: t=%f" % self.t)

    def get(self):
        return self.t

# setup (h)SPI for MAX7219 with dot LED matrix
#if False:
if True:
    # hardware SPI (boudrate up to 80 MHz, real about from 1 kHz to 13.25 MHz)
    spi = SPI(1, baudrate=10000000, polarity=0, phase=0, sck=Pin(14), mosi=Pin(13), miso=Pin(12))
else:
    # software SPI bus on the given pins (baudrate from 1 kHz to 500 kHz)
    # polarity is the idle state of SCK
    # phase=0 means sample on the first edge of SCK, phase=1 means the second
    spi = SPI(-1, baudrate=500000, polarity=0, phase=0, sck=Pin(14), mosi=Pin(13), miso=Pin(12))
    #spi.init(baudrate=100000) # set the baudrate

CS = 27 # CS on D27 (GPIO-27)
NUM = 4 # number of 8x8 dot matrix blocks

display = max7219.Matrix8x8(spi, Pin(CS, Pin.OUT), NUM)
display.brightness(1) # [0..15]
display.fill(0)
display.show()

"""
# put text (4x6 font) on frame buffer
def put_text(fbuf, text, x=0, y=0):
    for uchr in text:
        sym = FONT[ord(uchr)]
        for i in range(4):
            for j in range(6):
                fbuf.pixel(x + i, y + j,
                           1 if (sym[i] & (0x20 >> j)) else 0)
        x += 4
"""

# put text (6x8 font) on frame buffer
def put_text(fbuf, text, x=0, y=0):
    for uchr in text:
        sym = FONT[ord(uchr)]
        for i in range(6):
            for j in range(8):
                fbuf.pixel(x + i, y + j,
                           1 if (sym[i] & (0x80 >> j)) else 0)
        x += 6


# DHT22 on GPIO-15
dht = _DHT22_(15)

# DS18B20 on GPIO-4
ds = _DS18B20_(4)

# RTC based on DS3231
rtc = ds3231(sclPin=Pin(22), sdaPin=Pin(21))
#rtc.set(datetime=(2017, 12, 16, 0, 23, 40, 00))

# blink blue LED 3 times
led = Pin(2, Pin.OUT) # blue LED
for i in range(3):
    led.value(1)
    display.fill(1)
    display.show()
    time.sleep_ms(250)
    led.value(0)
    display.fill(0)
    display.show()
    time.sleep_ms(250)

put_text(display, "HELLO", x=1)
display.show()
time.sleep_ms(1000)


# FSM
cnt = 0
fsm = 0
#days = ("MO", "TU", "WE", "TH", "FR", "SA", "SU")
days = ("ПОНЕД", "ВТОРН", "СРЕДА", "ЧЕТВ", "ПЯТН", "СУББ", "ВОСКР")

def tick(timer):
    global dht, ds, rtc, cnt, fsm, days
    dht.run()
    ds.run()
    rtc.get()
    rtc.temp()

    # FSM
    cnt -= 1
    if cnt > 0:
        return

    # get data from sensors
    t1, h = dht.get()
    t2 = ds.get()
    dt, t3 = rtc.all()

    """
    if fsm == 1:
        fsm = 2 # show temperature
        cnt = 3
        txt = "T=%.1f°" % t2

    elif fsm == 2:
        fsm = 3 # showh humidity
        cnt = 2
        txt = "H=%.1f%%" % h
    
    else: # fsm == 3
        fsm = 1 # show time
        cnt = 5
        txt = "%2s %02i:%02i" % (days[dt[3] - 1], dt[4], dt[5])
    """ 

    if fsm == 1:
        fsm = 2 # show temperature
        cnt = 3
        txt = "%.1f°" % t2
    elif fsm == 2:
        fsm = 3 # showh humidity
        cnt = 2
        txt = "%.1f%%" % h
    elif fsm == 3:
        fsm = 4 # show date
        cnt = 2
        txt = "%02i.%02i" % (dt[1], dt[2])
    elif fsm == 4:
        fsm = 5 # show weekday
        cnt = 1
        txt = days[dt[3] - 1]
    else: # fsm == 5
        fsm = 1 # show time
        cnt = 6
        txt = "%02i:%02i" % (dt[4], dt[5])

    display.fill(0)
    put_text(display, txt, x=1, y=0)
    display.show()
    gc.collect()


# init periodic timer
timer = Timer(-1)
timer.init(period=1000, mode=Timer.PERIODIC, callback=tick)

"""
for i in range(2):
    dht.run()
    ds.run()
    rtc.get()
    rtc.temp()
"""

"""
# TCP server socket
addr = socket.getaddrinfo('0.0.0.0', 80)[0][-1]
s = socket.socket()
s.bind(addr)
s.listen(1)
                        
print('listening on', addr)
"""

# HTML template
html = """<!DOCTYPE html>
<html>
  <head><title>ESP32 &#043; DHT22 &#043; DS18B20 &#043; DS3231</title> </head>
  <body>
    <h1>DHT22:</h1>
    <h2>Temperature = %s &deg;C</h2>
    <h2>Humidity = %s &#037;</h2>
    <hr/>
    <h1>DS18B20:</h1>
    <h2>Temperature = %s &deg;C</h2>
    <hr/>
    <h1>DS3231:</h1>
    <h2>DateTime = %s</h2>
    <h2>Temperature = %s &deg;C</h2>
  </body>
</html>
"""

gc.collect()

"""
# Web server
while True:
    cl, addr = s.accept()
    #print('client connected from', addr)

    cl_file = cl.makefile('rwb', 0)
    while True:
        line = cl_file.readline()
        if not line or line == b'\r\n':
            break
    
    led.value(1)
    t1, h = dht.get()
    t2 = ds.get()
    dt, t3 = rtc.all()

    # format date and time to YYYY.MM.DD HH:MM:SS
    sdt = "%4i.%02i.%02i %02i:%02i:%02i" % \
          (dt[0], dt[1], dt[2], dt[4], dt[5], dt[6])

    response = html % (str(t1), str(h), str(t2), sdt, str(t3))
    cl.send(response)
    cl.close()

    led.value(0)
    gc.collect()
"""




max7219.py
# MicroPython max7219 8x8 LED matrix driver, cascadable, SPI interface

from micropython import const
import framebuf

_NOOP = const(0)
_DIGIT0 = const(1)
_DECODEMODE = const(9)
_INTENSITY = const(10)
_SCANLIMIT = const(11)
_SHUTDOWN = const(12)
_DISPLAYTEST = const(15)

class Matrix8x8:
    def __init__(self, spi, cs, num):
        """
        Driver for cascading MAX7219 8x8 LED matrices.

        >>> import max7219
        >>> from machine import Pin, SPI
        >>> spi = SPI(1)
        >>> display = max7219.Matrix8x8(spi, Pin('X5'), 4)
        >>> display.text('1234',0,0,1)
        >>> display.show()

        """
        self.spi = spi
        self.cs = cs
        self.cs.init(cs.OUT, True)
        self.buffer = bytearray(8 * num)
        self.num = num
        fb = framebuf.FrameBuffer(self.buffer, 8 * num, 8, framebuf.MONO_HLSB)
        self.framebuf = fb
        # Provide methods for accessing FrameBuffer graphics primitives. This is a workround
        # because inheritance from a native class is currently unsupported.
        # http://docs.micropython.org/en/latest/pyboard/library/framebuf.html
        self.fill = fb.fill  # (col)
        self.pixel = fb.pixel # (x, y[, c])
        self.hline = fb.hline  # (x, y, w, col)
        self.vline = fb.vline  # (x, y, h, col)
        self.line = fb.line  # (x1, y1, x2, y2, col)
        self.rect = fb.rect  # (x, y, w, h, col)
        self.fill_rect = fb.fill_rect  # (x, y, w, h, col)
        self.text = fb.text  # (string, x, y, col=1)
        self.scroll = fb.scroll  # (dx, dy)
        self.blit = fb.blit  # (fbuf, x, y[, key])
        self.init()

    def _write(self, command, data):
        self.cs(0)
        for m in range(self.num):
            #!!!self.spi.write(bytearray([command, data]))
            buf = bytearray([command, data])
            self.spi.write_readinto(buf, buf)
        self.cs(1)

    def init(self):
        for command, data in (
            (_SHUTDOWN, 0),
            (_DISPLAYTEST, 0),
            (_SCANLIMIT, 7),
            (_DECODEMODE, 0),
            (_SHUTDOWN, 1),
        ):
            self._write(command, data)

    def brightness(self, value):
        if not 0 <= value <= 15:
            raise ValueError("Brightness out of range")
        self._write(_INTENSITY, value)

    def show(self):
        for y in range(8):
            self.cs(0)
            for m in range(self.num):
                #!!!self.spi.write(bytearray([_DIGIT0 + y, self.buffer[(y * self.num) + m]]))
                buf = bytearray([_DIGIT0 + y, self.buffer[(y * self.num) + m]])
                self.spi.write_readinto(buf, buf)
            self.cs(1)



font_4x6.py
font_4x6 = {
  0x0020: b"\x00\x00\x00\x00", # ' '
  0x0021: b"\x00\x1D\x00\x00", # '!'
  0x0022: b"\x18\x00\x18\x00", # '"'
  0x0023: b"\x0E\x1B\x0E\x00", # '#'
  0x0024: b"\x0D\x1F\x16\x00", # '$'
  0x0025: b"\x12\x04\x09\x00", # '%'
  0x0026: b"\x0A\x15\x0B\x00", # '&'
  0x0027: b"\x00\x18\x00\x00", # '''
  0x0028: b"\x0E\x11\x00\x00", # '('
  0x0029: b"\x00\x11\x0E\x00", # ')'
  0x002A: b"\x0A\x04\x0A\x00", # '*'
  0x002B: b"\x04\x0E\x04\x00", # '+'
  0x002C: b"\x01\x02\x00\x00", # ','
  0x002D: b"\x04\x04\x04\x00", # '-'
  0x002E: b"\x00\x01\x00\x00", # '.'
  0x002F: b"\x02\x04\x08\x00", # '/'
  0x0030: b"\x1F\x11\x1F\x00", # '0'
  0x0031: b"\x00\x08\x1F\x00", # '1'
  0x0032: b"\x17\x15\x1D\x00", # '2'
  0x0033: b"\x15\x15\x1F\x00", # '3'
  0x0034: b"\x1C\x04\x1F\x00", # '4'
  0x0035: b"\x1D\x15\x17\x00", # '5'
  0x0036: b"\x1F\x15\x17\x00", # '6'
  0x0037: b"\x10\x10\x1F\x00", # '7'
  0x0038: b"\x1F\x15\x1F\x00", # '8'
  0x0039: b"\x1D\x15\x1F\x00", # '9'
  0x003A: b"\x00\x0A\x00\x00", # ':'
  0x003B: b"\x01\x0A\x00\x00", # ';'
  0x003C: b"\x04\x0A\x11\x00", # '<'
  0x003D: b"\x0A\x0A\x0A\x00", # '='
  0x003E: b"\x11\x0A\x04\x00", # '>'
  0x003F: b"\x18\x13\x1C\x00", # '?'
  0x0040: b"\x1C\x14\x1C\x00", # '@'
  0x0041: b"\x0F\x14\x1F\x00", # 'A'
  0x0042: b"\x1F\x15\x0A\x00", # 'B'
  0x0043: b"\x0E\x11\x11\x00", # 'C'
  0x0044: b"\x1F\x11\x0E\x00", # 'D'
  0x0045: b"\x1F\x15\x15\x00", # 'E'
  0x0046: b"\x1F\x14\x14\x00", # 'F'
  0x0047: b"\x0E\x11\x13\x00", # 'G'
  0x0048: b"\x1F\x04\x1F\x00", # 'H'
  0x0049: b"\x11\x1F\x11\x00", # 'I'
  0x004A: b"\x11\x1E\x10\x00", # 'J'
  0x004B: b"\x1F\x04\x1B\x00", # 'K'
  0x004C: b"\x1F\x01\x01\x00", # 'L'
  0x004D: b"\x1F\x0C\x1F\x00", # 'M'
  0x004E: b"\x1F\x10\x0F\x00", # 'N'
  0x004F: b"\x1F\x11\x1F\x00", # 'O'
  0x0050: b"\x1F\x14\x1C\x00", # 'P'
  0x0051: b"\x0E\x11\x0F\x00", # 'Q'
  0x0052: b"\x1F\x14\x0B\x00", # 'R'
  0x0053: b"\x09\x15\x12\x00", # 'S'
  0x0054: b"\x10\x1F\x10\x00", # 'T'
  0x0055: b"\x1E\x01\x1F\x00", # 'U'
  0x0056: b"\x1E\x01\x1E\x00", # 'V'
  0x0057: b"\x1F\x06\x1F\x00", # 'W'
  0x0058: b"\x1B\x04\x1B\x00", # 'X'
  0x0059: b"\x18\x07\x18\x00", # 'Y'
  0x005A: b"\x13\x15\x19\x00", # 'Z'
  0x005B: b"\x1F\x11\x00\x00", # '['
  0x005C: b"\x08\x04\x02\x00", # '\'
  0x005D: b"\x00\x11\x1F\x00", # ']'
  0x005E: b"\x08\x10\x08\x00", # '^'
  0x005F: b"\x01\x01\x01\x00", # '_'
  0x0060: b"\x00\x10\x08\x00", # '`'
  0x0061: b"\x12\x15\x0F\x00", # 'a'
  0x0062: b"\x1F\x05\x02\x00", # 'b'
  0x0063: b"\x02\x05\x05\x00", # 'c'
  0x0064: b"\x02\x05\x1F\x00", # 'd'
  0x0065: b"\x06\x0B\x05\x00", # 'e'
  0x0066: b"\x00\x0F\x14\x00", # 'f'
  0x0067: b"\x08\x15\x1E\x00", # 'g'
  0x0068: b"\x1F\x04\x03\x00", # 'h'
  0x0069: b"\x00\x16\x01\x00", # 'i'
  0x006A: b"\x02\x01\x16\x00", # 'j'
  0x006B: b"\x1F\x04\x0B\x00", # 'k'
  0x006C: b"\x00\x1E\x01\x00", # 'l'
  0x006D: b"\x07\x06\x03\x00", # 'm'
  0x006E: b"\x07\x04\x03\x00", # 'n'
  0x006F: b"\x07\x05\x07\x00", # 'o'
  0x0070: b"\x0F\x0A\x0E\x00", # 'p'
  0x0071: b"\x04\x0A\x0F\x00", # 'q'
  0x0072: b"\x03\x04\x04\x00", # 'r'
  0x0073: b"\x0D\x0B\x00\x00", # 's'
  0x0074: b"\x08\x1E\x09\x00", # 't'
  0x0075: b"\x06\x01\x07\x00", # 'u'
  0x0076: b"\x06\x01\x06\x00", # 'v'
  0x0077: b"\x07\x03\x07\x00", # 'w'
  0x0078: b"\x05\x02\x05\x00", # 'x'
  0x0079: b"\x04\x03\x04\x00", # 'y'
  0x007A: b"\x0B\x0D\x09\x00", # 'z'
  0x007B: b"\x04\x1B\x11\x00", # '{'
  0x007C: b"\x00\x1F\x00\x00", # '|'
  0x007D: b"\x11\x1B\x04\x00", # '}'
  0x007E: b"\x08\x18\x10\x00", # '~'
  0x00B0: b"\x08\x14\x08\x00", # '°'
  0x00B4: b"\x00\x08\x10\x00", # '´'
  0x00B7: b"\x00\x04\x00\x00", # '·'
}



ds3231.py
# MicroPython DS3231 RTC driver, I2C interface for ESP8266/ESP32

from machine import Pin, I2C
import ucollections
import utime

"""
 ESP32     | DS3231 for Pi
 ----------+--------------
 3.3       | 1 - VCC (+)
 D21 (SDA) | 2 - SDA
 D22 (SCL) | 3 - SCL
           | 4 - NC
 GND       | 5 - GND
"""

DateTimeTuple = ucollections.namedtuple("DateTimeTuple",
  ["year", "month", "day", "weekday", "hour", "minute", "second", "millisecond"])

def datetime_tuple(year, month, day, weekday=0, hour=0, minute=0,
                   second=0, millisecond=0):
  return DateTimeTuple(year, month, day, weekday, hour, minute,
                       second, millisecond)

def bcd2bin(value):
  return value - 6 * (value >> 4)

def bin2bcd(value):
    return value + 6 * (value // 10)

def tuple2seconds(datetime):
  return utime.mktime((datetime.year, datetime.month, datetime.day,
    datetime.hour, datetime.minute, datetime.second, datetime.weekday, 0))

def seconds2tuple(seconds):
  year, month, day, hour, minute, second, weekday, _yday = utime.localtime()
  return DateTimeTuple(year, month, day, weekday, hour, minute, second, 0)

class ds3231:
  DATETIME_REGISTER    = 0x00
  TEMPERATURE_REGISTER = 0x11  

  def __init__(self, sclPin=Pin(22), sdaPin=Pin(21), address=104):
    self.address = address
    self.i2c = I2C(scl=sclPin, sda=sdaPin, freq=100000)
    scan = self.i2c.scan()
    if address not in scan:
      print("error: DS3231 not found on I2C bus")
    print("i2c.scan():", scan)
    self.get()
    self.temp()


  def get(self):
    buf = bytearray(7)
    buf = self.i2c.readfrom_mem(self.address, self.DATETIME_REGISTER, 7)
    #print(buf) # FIXME
    self.dt = datetime_tuple(
                year=bcd2bin(buf[6]) + 2000,
                month=bcd2bin(buf[5]),
                day=bcd2bin(buf[4]),
                weekday=bcd2bin(buf[3]),
                hour=bcd2bin(buf[2]),
                minute=bcd2bin(buf[1]),
                second=bcd2bin(buf[0]),
            )
    return self.dt


  def set(self, datetime=(2018, 12, 31, 0, 23, 59, 59, 999)):
    buf = bytearray(7)
    datetime = datetime_tuple(*datetime)
    buf[0] = bin2bcd(datetime.second)
    buf[1] = bin2bcd(datetime.minute)
    buf[2] = bin2bcd(datetime.hour)
    buf[3] = bin2bcd(datetime.weekday)
    buf[4] = bin2bcd(datetime.day)
    buf[5] = bin2bcd(datetime.month)
    buf[6] = bin2bcd(datetime.year - 2000)
    retv = self.i2c.writeto_mem(self.address, self.DATETIME_REGISTER, buf)
    #print(buf)  # FIXME
    #print(retv) # FIXME
    self.get()


  def temp(self):
    buf = self.i2c.readfrom_mem(self.address, self.TEMPERATURE_REGISTER, 2)
    #print(buf[0], buf[1] >> 6) # FIXME
    temp = buf[0] + (buf[1] >> 6) * 0.25
    if temp > 127.:
      temp -= 256. # FIXME
    self.t = temp
    return temp

  def all(self):
      return self.dt, self.t

